File: typescript/src/crc32.test.ts
/**
 * Tests for the CRC32 checksum module.
 *
 * CRC32 is used to verify data integrity in the BySquare format.
 * Tests verify correct checksum calculation for various inputs.
 */

import {
	describe,
	expect,
	test,
} from "bun:test";

import { crc32 } from "./crc32.js";

/**
 * Test cases with known CRC32 values.
 * Values verified against standard CRC32 implementations.
 */
const crc32TestCases = [
	{ name: "empty string", input: "", expected: 0 },
	{ name: "single character", input: "a", expected: 3904355907 },
	{ name: "hello world", input: "hello world", expected: 222957957 },
	{ name: "123456789", input: "123456789", expected: 3421780262 },
];

describe("crc32", () => {
	test.each(crc32TestCases)("$name", ({ input, expected }) => {
		const result = crc32(input);
		expect(result).toBe(expected);
	});
});

File: go/pkg/bysquare/crc32_test.go
package bysquare

import "testing"

func TestCRC32Checksum(t *testing.T) {
	testCases := []struct {
		name     string
		input    string
		expected uint32
	}{
		{
			name:     "empty string",
			input:    "",
			expected: 0,
		},
		{
			name:     "single character",
			input:    "a",
			expected: 3904355907,
		},
		{
			name:     "hello world",
			input:    "hello world",
			expected: 222957957,
		},
		{
			name:     "numeric string",
			input:    "123456789",
			expected: 3421780262,
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result := CRC32(tc.input)
			if result != tc.expected {
				t.Errorf("expected %d, got %d", tc.expected, result)
			}
		})
	}
}

func TestCRC32ChecksumConsistency(t *testing.T) {
	input := "test data for consistency"
	first := CRC32(input)
	second := CRC32(input)

	if first != second {
		t.Errorf("CRC32 checksum not consistent: first=%d, second=%d", first, second)
	}
}

func TestCRC32ChecksumDifferentInputs(t *testing.T) {
	input1 := "test"
	input2 := "test2"

	checksum1 := CRC32(input1)
	checksum2 := CRC32(input2)

	if checksum1 == checksum2 {
		t.Error("different inputs should produce different checksums")
	}
}
